cardano-api/src/Cardano/Api/Tx.hs

```diff
@@ -391,61 +400,45 @@ instance IsCardanoEra era => SerialiseAsCBOR (KeyWitness era) where
     serialiseToCBOR (ByronKeyWitness wit) = CBOR.serialize' wit
 
     serialiseToCBOR (ShelleyKeyWitness _era wit) =
-      CBOR.serializeEncoding' $
-      encodeShelleyBasedKeyWitness wit
+      CBOR.serialize' wit
 
     serialiseToCBOR (ShelleyBootstrapWitness _era wit) =
-      CBOR.serializeEncoding' $
-      encodeShelleyBasedBootstrapWitness wit
+      CBOR.serialize' wit
 
     deserialiseFromCBOR _ bs =
       case cardanoEra :: CardanoEra era of
-        ByronEra ->
-          ByronKeyWitness <$> CBOR.decodeFull' bs
-
-        -- Use the same derialisation impl, but at different types:
-        ShelleyEra -> decodeShelleyBasedWitness ShelleyBasedEraShelley bs
-        AllegraEra -> decodeShelleyBasedWitness ShelleyBasedEraAllegra bs
-        MaryEra    -> decodeShelleyBasedWitness ShelleyBasedEraMary    bs
-        AlonzoEra  -> decodeShelleyBasedWitness ShelleyBasedEraAlonzo  bs
-        BabbageEra -> decodeShelleyBasedWitness ShelleyBasedEraBabbage bs
-
-
-encodeShelleyBasedKeyWitness :: ToCBOR w => w -> CBOR.Encoding
-encodeShelleyBasedKeyWitness wit =
-    CBOR.encodeListLen 2 <> CBOR.encodeWord 0 <> toCBOR wit
-
-encodeShelleyBasedBootstrapWitness :: ToCBOR w => w -> CBOR.Encoding
-encodeShelleyBasedBootstrapWitness wit =
-    CBOR.encodeListLen 2 <> CBOR.encodeWord 1 <> toCBOR wit
-
-decodeShelleyBasedWitness :: forall era.
-                             ShelleyBasedEra era
-                          -> ByteString
-                          -> Either CBOR.DecoderError (KeyWitness era)
-decodeShelleyBasedWitness era =
-    CBOR.decodeAnnotator "Shelley Witness" decode . LBS.fromStrict
-  where
-    decode :: CBOR.Decoder s (CBOR.Annotator (KeyWitness era))
-    decode =  do
-      CBOR.decodeListLenOf 2
-      t <- CBOR.decodeWord
-      case t of
-        0 -> fmap (fmap (ShelleyKeyWitness era)) fromCBOR
-        1 -> fmap (fmap (ShelleyBootstrapWitness era)) fromCBOR
-        _ -> CBOR.cborError $ CBOR.DecoderErrorUnknownTag
-                                "Shelley Witness" (fromIntegral t)
+        ByronEra -> ByronKeyWitness <$> CBOR.decodeFull' bs
+        ShelleyEra -> decodeKeyOrBootstrapWitness ShelleyBasedEraShelley bs
+        AllegraEra -> decodeKeyOrBootstrapWitness ShelleyBasedEraAllegra bs
+        MaryEra -> decodeKeyOrBootstrapWitness ShelleyBasedEraMary bs
+        AlonzoEra -> decodeKeyOrBootstrapWitness ShelleyBasedEraAlonzo bs
+        BabbageEra -> decodeKeyOrBootstrapWitness ShelleyBasedEraBabbage bs
+
+decodeKeyOrBootstrapWitness
+  :: ShelleyBasedEra era -> ByteString ->  Either CBOR.DecoderError (KeyWitness era)
+decodeKeyOrBootstrapWitness era bs =
+  case CBOR.decodeAnnotator "KeyWitness" fromCBOR (LBS.fromStrict bs) of
+    Right keyWit -> return $ ShelleyKeyWitness era keyWit
+    Left{} ->
```

Comment:
What's the reason for ignoring `Left` payload?

Reply:
So that we can try to deserialise a bootstrap witness